### PR TITLE
Topic/device naming

### PR DESCRIPTION
**Pull request by bosilca**
PARSEC_HAVE_CUDA is for CMake. In PaRSEC we should use PARSEC_HAVE_DEV_CUDA_SUPPORT. This applies to all devices, including RECURSIVE and CPU. 

This PR also allows for specifying [type=CPU] next to a body and still do the right thing.

### CODE REVIEW

**Comment by abouteiller** [comment]
Looking good but I need to double check it still work on HIP I remember there was more than aesthetic reasons for the lowercasing of the hook.

**Review by QingleiCao — APPROVED** [review]

File: parsec/include/parsec/parsec_options.h.in [hunk 1]
```diff
@@ -120,17 +120,28 @@
...
 #cmakedefine PARSEC_HAVE_MPI_30
 #cmakedefine PARSEC_HAVE_MPI_OVERTAKE
 #cmakedefine PARSEC_HAVE_AYUDAME
 
+#cmakedefine PARSEC_HAVE_DEV_CPU_SUPPORT
+#cmakedefine PARSEC_HAVE_DEV_RECURSIVE_SUPPORT
+#cmakedefine PARSEC_HAVE_DEV_CUDA_SUPPORT
+#cmakedefine PARSEC_HAVE_DEV_HIP_SUPPORT
+#cmakedefine PARSEC_HAVE_DEV_OPENCL_SUPPORT
+
```
[the remark this answers is not in the document]
Comment (bosilca) [reply parsec/include/parsec/parsec_options.h.in:144]: the PARSEC_HAVE_DEV* is for C code purposes and indicates what devices are supported by a particular build of parsec, while PARSEC_HAVE_* indicates that the CMake found support for the specific functionality. Moreover there is a directional dependency between then as PARSEC_HAVE_DEV_CUDA_SUPPORT will imply that PARSEC_HAVE_CUDA is also enabled, but the reciprocal is not true.

File: parsec/interfaces/ptg/ptg-compiler/jdf2c.c [hunk 2]
```diff
@@ -3918,14 +3918,16 @@ jdf_generate_function_incarnation_list( const jdf_t *jdf,
...
         jdf_find_property(body->properties, "dyld", &dyld_property);
         if( NULL == type_property) {
+            string_arena_add_string(sa, "#if defined(PARSEC_HAVE_DEV_CPU_SUPPORT)\n");
             string_arena_add_string(sa, "    { .type     = PARSEC_DEV_CPU,\n");
             string_arena_add_string(sa, "      .evaluate = %s,\n", "NULL");
             string_arena_add_string(sa, "      .hook     = (parsec_hook_t*)hook_of_%s },\n", base_name);
+            string_arena_add_string(sa, "#endif  /* defined(PARSEC_HAVE_DEV_CPU_SUPPORT) */\n");
         } else {
             char* dev_upper = strdup_upper(type_property->expr->jdf_var);
             char* dev_lower = strdup_lower(type_property->expr->jdf_var);
```
[the remark this answers is not in the document]
Comment (bosilca) [reply parsec/interfaces/ptg/ptg-compiler/jdf2c.c]: gone.

File: tests/dsl/ptg/recursive.jdf [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+/*
+ * A function to recursively complete a recursive taskpool.
+ */
+static void parsec_recursive_callback(parsec_taskpool_t* _tp, const parsec_recursive_callback_t* data)
+{
+    parsec_recursive_taskpool_t* tp = (parsec_recursive_taskpool_t*)_tp;
+    __parsec_recursive_DO_SOMETHING_task_t* task = (__parsec_recursive_DO_SOMETHING_task_t*)data->task;
+
+    /*parsec_taskpool_free(tp);*/
+    PARSEC_OBJ_RELEASE(_tp);
```
[the remark this answers is not in the document]
Comment (bosilca) [reply tests/dsl/ptg/recursive.jdf:34]: I don't know, I need to take another look at this now the termination detection PR has been merged. At the end of the day I think that `parsec_taskpool_free` should be just a way for the user to relinquish it's reference to the object, so it should fall back to calling `PARSEC_OBJ_RELEASE`.

**Review by therault — CHANGES_REQUESTED** [review]
`parsec/mca/device/cuda/ValidateModule.cmake` still refers to `if( PARSEC_HAVE_CUDA )
`. If I understand the logic, it should do `if( PARSEC_HAV_DEV_CUDA_SUPPORT )` and the module should not be compiled at all if the user deactivates CUDA supports via the CMake option?

**Comment by bosilca** [comment]
I am not sure I understand the comment. The PARSEC_HAVE_CUDA is only for the CMake level logic, and this includes ValidateModule.cmake. We are not supposed to use it in the code, where we should use `PARSEC_HAVE_DEV_CUDA_SUPPORT` (and CUDA can be replaced with any other device name all in capitals).

File: parsec/include/parsec/parsec_options.h.in [hunk 4]
```diff
@@ -120,17 +120,28 @@
 #define CMAKE_PARSEC_C_FLAGS      "@CMAKE_C_FLAGS@"
 #define CMAKE_PARSEC_C_INCLUDES   "@PARSEC_C_INCLUDES@"
 
-#cmakedefine PARSEC_HAVE_HWLOC
-#cmakedefine PARSEC_HAVE_PAPI
+/**
+ * The following 3 defines are deprecated. Instead users should
+ * rely on the corresponding PARSEC_HAVE_DEV_XXX_SUPPORT
+ */
 #cmakedefine PARSEC_HAVE_CUDA
```
Comment (bosilca) [inline parsec/include/parsec/parsec_options.h.in:127]: remove these 3 variables they should remain at the CMake level.

**Review by therault — APPROVED** [review]

**Review by abouteiller — APPROVED** [review]

File: parsec/include/parsec/parsec_options.h.in (hunk 4 above)
Comment (abouteiller) [reply parsec/include/parsec/parsec_options.h.in:127]: do you want to do this before we merge?

**Comment by abouteiller** [comment]
ctest is busted on test `stage`

**Pull request merged**